luigi/worker.py

```diff
@@ -737,28 +782,35 @@ def add(self, task, multiprocess=False, processes=0):
             self._first_task = task.task_id
         self.add_succeeded = True
         if multiprocess:
-            queue = multiprocessing.Manager().Queue()
             pool = multiprocessing.Pool(processes=processes if processes > 0 else None)
         else:
-            queue = DequeQueue()
             pool = SingleProcessPool()
         self._validate_task(task)
-        pool.apply_async(check_complete, [task, queue])
+        results = [(task, pool.apply_async(check_complete, (task,)))]
 
-        # we track queue size ourselves because len(queue) won't work for multiprocessing
-        queue_size = 1
         try:
             seen = {task.task_id}
-            while queue_size:
-                current = queue.get()
-                queue_size -= 1
-                item, is_complete = current
-                for next in self._add(item, is_complete):
-                    if next.task_id not in seen:
-                        self._validate_task(next)
-                        seen.add(next.task_id)
-                        pool.apply_async(check_complete, [next, queue])
-                        queue_size += 1
+            while results:
+                # fetch the first done result
+                for i, (task, result) in enumerate(list(results)):
+                    if result.ready():
+                        results.pop(i)
+                        break
+                else:
+                    time.sleep(wait_interval)
+                    continue
```

Comment:
The for/else + continue makes the flow unclear. How about packing up the rest of the loop into a new method and calling it before `break`?